### PR TITLE
Update Traefik according to api-platform V2.4 docker-compose declaration

### PR DESCRIPTION
**Pull request by Darkweak**
Implements  into traefik doc file
* mercure and admin declaration
* going further to define all in .env file

### CODE REVIEW

File: deployment/traefik.md [hunk 1]
```diff
@@ -12,29 +12,38 @@ Override ports and add labels to tell Traefik to listen on the routes mentionned
...
-      - "traefik.frontend.rule=Host:api.localhost"
+      - traefik.frontend.rule=Host:api.localhost
 ``` 
 The API DNS will be specified with `traefik.frontend.rule=Host:your.host` (here api.localhost)  
 
 `--traefik.port=3000` The port specified to Traefik will be exposed by the container (here the React app exposes the 3000 port)  
 
 
 ```yaml
+# api-platform-app/docker-compose.yml
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
# docker-compose.yml
```

File: deployment/traefik.md [hunk 2]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
 ## Known Issues
 
 If your network is of type B, it may conflict with the Traefik sub-network.
+
+## Going Further
+
+You may want multiple api-platform instances on same server. To deal with it, you'll have to externalize first traefik to another docker-compose file anywhere on your server, but as traefik will listen on external 80 & 443 ports you can run only 1 traefik instance with this configuration per server
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
As this Traefik configuration listens on 80 and 443 ports, you can run only 1 Traefik instance per server. However, you may want to run multiple API Platform projects on same server. To deal with it, you'll have to externalize the Traefik configuration to another `docker-compose.yml` file, anywhere on your server.
```

File: deployment/traefik.md [hunk 3]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
 ## Known Issues
 
 If your network is of type B, it may conflict with the Traefik sub-network.
+
+## Going Further
+
+You may want multiple api-platform instances on same server. To deal with it, you'll have to externalize first traefik to another docker-compose file anywhere on your server, but as traefik will listen on external 80 & 443 ports you can run only 1 traefik instance with this configuration per server  
+Then you'll have something like this example :
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
Here is a working example:  
```

File: deployment/traefik.md [hunk 4]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
 ## Known Issues
 
 If your network is of type B, it may conflict with the Traefik sub-network.
+
+## Going Further
+
+You may want multiple api-platform instances on same server. To deal with it, you'll have to externalize first traefik to another docker-compose file anywhere on your server, but as traefik will listen on external 80 & 443 ports you can run only 1 traefik instance with this configuration per server  
+Then you'll have something like this example :  
+```yaml
+# traefik/docker-compose.yml
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
# /somewhere/docker-compose.yml
```

File: deployment/traefik.md [hunk 5]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
...
+  traefik:
+    image: traefik
+    command: --api --docker
+    ports:
+      - "80:80"
+      - "443:443"
+      - "8080:8080"
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock
+# Uncomment these following lines to enable toml files & let's encrypt certificates, refer to traefik doc : https://docs.traefik.io/user-guide/docker-and-lets-encrypt/
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
# Uncomment these following lines to load a TOML configuration file and to generate Let's Encrypt certificated as explained in https://docs.traefik.io/user-guide/docker-and-lets-encrypt/
```

File: deployment/traefik.md [hunk 6]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
...
+      - api_platform_network
+    # Add other networks here
+
+networks:
+  api_platform_network:
+    external: true
+  # Add other networks here
+```
+
+After that, you must define your api-platform docker-compose file
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
Then update the `docker-compose.yaml` file belonging to your API Platform projects:

```

File: deployment/traefik.md [hunk 7]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
...
+
+networks:
+  api_platform_network:
+    external: true
+  # Add other networks here
+```
+
+After that, you must define your api-platform docker-compose file  
+```yaml
+# api-platform-app/docker-compose.yml
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
# docker-compose.yml
```

File: deployment/traefik.md [hunk 8]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
...
+version: '3.4'
```
Comment (dunglas) [inline deployment/traefik.md]: I suggest to use the patch format instead:

```
```patch
# docker-compose.yml
-  traefik:
-    image: traefik:latest
-    command: --api --docker
-    ports:
-      - "80:80"
-      - "443:443"
-    volumes:
-      - /var/run/docker.sock:/var/run/docker.sock
-    <<: *network 
```
```

File: deployment/traefik.md [hunk 9]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
...
+
+volumes:
+  db-data: {}
+
+networks:
+  api_platform_network:
+    external: true
+```
+
+Then as you can see, there is some env vars must be defined, here is an example of .env file :
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
Finally, some environment variables must be defined, here is an example of a `.env` file to set them:  
```

File: deployment/traefik.md [hunk 10]
```diff
@@ -120,3 +178,195 @@ If you do that, you'll have to update the `CORS_ALLOW_ORIGIN` environment variab
...
+DOMAIN_NAME=localhost
+HTTP_OR_SSL=http://
+DB_NAME=api-platform-db-name
+DB_PASS=YouMustChangeThisPassword
+DB_USER=api-platform
+JWT_KEY=!UnsecureChangeMe!
+SUBDOMAINS_LIST=(admin|api|cache|mercure|www)
+```
+
+This way let you configure your main variables into one single file
```
Comment (dunglas) [inline deployment/traefik.md]:
```suggestion
This way, you can configure your main variables into one single file.
```

**Comment by Darkweak** [comment]
ping @dunglas Fixes are done

**Comment by teohhanhui** [comment]
Is there a reason why we don't just use Traefik in api-platform/api-platform distribution? Or does it add much complexity?

**Comment by Darkweak** [comment]
@dunglas told me it wasn't necessary to including this on distribution, so it's just in doc part

**Comment by teohhanhui** [comment]
I think it'd be desirable to include Traefik out of the box in the distribution. WDYT @dunglas?

But I'm not sure how feasible it is to have the hostnames be resolved correctly, out of the box.

**Comment by teohhanhui** [comment]
If only Docker supports zeroconf: https://en.wikipedia.org/wiki/Zero-configuration_networking 😂

**Comment by teohhanhui** [comment]
https://github.com/containous/traefik/issues/4092

**Comment by dunglas** [comment]
Traefik introduces a one more dependency, and some complexity. Also, from my experience, it doesn't play very well with Kubernetes (it's supported, but it's a bit hacky, especially with Helm). I would prefer to keep Traefik setup as a docs entry for now

**Comment by Darkweak** [comment]
So, just merge this, at the moment we won't implement it into API P

**Comment by dunglas** [comment]
Thanks @Darkweak

**Pull request merged**